### PR TITLE
Cron schedule for dumping in external mongo

### PR DESCRIPTION
**Pull request by HughParry**

### CODE REVIEW

File: packages/database/src/captchaDatabase/captchaDatabse.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { UserCommitmentRecord } from '@prosopo/types-database'
+import { getLoggerDefault } from '@prosopo/common'
+import mongoose from 'mongoose'
+const logger = getLoggerDefault()
+
+const storedCaptchaSchema = new mongoose.Schema({
```
Comment (forgetso) [inline packages/database/src/captchaDatabase/captchaDatabse.ts]: Use `UserCommitmentRecordSchema` from `captcha/packages/types-database/src/types/mongo.ts`

File: packages/database/src/databases/mongo.ts [hunk 2]
```diff
@@ -565,6 +565,23 @@ export class ProsopoDatabase extends AsyncFactory implements Database {
...
+        const docs = await this.tables?.commitment.find({ stored: false }).lean()
+        return docs ? docs.map((doc) => UserCommitmentSchema.parse(doc)) : []
+    }
+
+    /** @description Mark a list of captcha commits as stored
+     */
+    async markDappUserCommitmentsStored(commitmentIds: Hash[]): Promise<void> {
+        await this.tables?.commitment.updateMany(
+            { id: { $in: commitmentIds } },
+            { $set: { stored: true } },
```
Comment (forgetso) [inline packages/database/src/databases/mongo.ts:584]: Does `stored` need to be added to `UserCommitmentRecordSchema` and `UserCommitmentRecord`?

File: packages/provider/src/batch/mongoBatch.ts
Comment (forgetso) [inline packages/provider/src/batch/mongoBatch.ts:1]: What is this?

File: packages/provider/src/captchaScheduler.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { CronJob } from 'cron'
+import { KeyringPair } from '@polkadot/keyring/types'
+import { ProsopoConfigOutput } from '@prosopo/types'
+import { ProsopoEnvError } from '@prosopo/common'
+import { ProviderEnvironment } from '@prosopo/env'
+import { Tasks } from './tasks/tasks.js'
+
+export async function batchCommitScheduler(pair: KeyringPair, config: ProsopoConfigOutput) {
```
Comment (forgetso) [inline packages/provider/src/captchaScheduler.ts]: Calling this batch commit is pretty confusing considering the old batch commit to contract stuff that's kicking about. How about `storeCaptchasExternally`?

Is the contents of this file meant to be in the empty file above?

File: packages/provider/src/captchaScheduler.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const cronSchedule = '0 * * * *'
+
+    const job = new CronJob(cronSchedule, async () => {
+        env.logger.debug('storeCommitmentsExternal task....')
+        await tasks.storeCommitmentsExternal().catch((err) => {
+            env.logger.error(err)
+        })
+    })
+
+    job.start()
```
Comment (forgetso) [inline packages/provider/src/captchaScheduler.ts]: How is this scheduler started?

**Review by forgetso — CHANGES_REQUESTED** [review]

File: packages/database/src/captchaDatabase/captchaDatabse.ts (hunk 1 above)
Comment (HughParry) [reply packages/database/src/captchaDatabase/captchaDatabse.ts]: Done

File: packages/database/src/databases/mongo.ts (hunk 2 above)
Comment (HughParry) [reply packages/database/src/databases/mongo.ts:584]: Yeah it's the most straightforward way to make sure that records are only added to mongo once

Comment (forgetso) [reply packages/database/src/databases/mongo.ts:584]: Sure, but `stored` is missing from the types. I think it needs to be added to `UserCommitmentRecordSchema` and `UserCommitmentRecord`

File: packages/provider/src/captchaScheduler.ts (hunk 3 above)
Comment (HughParry) [reply packages/provider/src/captchaScheduler.ts]: Changed

File: packages/provider/src/captchaScheduler.ts (hunk 4 above)
Comment (HughParry) [reply packages/provider/src/captchaScheduler.ts]: Forgot to add to cli, added now

File: packages/database/src/databases/mongo.ts (hunk 2 above)
Comment (HughParry) [reply packages/database/src/databases/mongo.ts:584]: Ah gotcha, just added

File: packages/provider/src/captchaScheduler.ts (hunk 4 above)
Comment (HughParry) [reply packages/provider/src/captchaScheduler.ts]: Resolved since added now

**Review by forgetso — CHANGES_REQUESTED** [review]
Close the database connection and we're good to go

**Review by forgetso — APPROVED** [review]

**Pull request merged**